caikit/runtime/health_probe.py

```diff
@@ -0,0 +1,267 @@
+                    f"{protocol}://", requests.adapters.HTTPAdapter(max_retries=retries)
+                )
+                resp = session.get(
+                    f"{protocol}://localhost:{port}{HEALTH_ENDPOINT}",
+                    timeout=0.01,
+                    **kwargs,
+                )
+            resp.raise_for_status()
+            return True
+        except requests.exceptions.SSLError as err:
+            log.debug("Got SSLError indicating a healthy SSL server! %s", err)
+            log.debug2(err, exc_info=True)
+            return True
+        except Exception as err:  # pylint: disable=broad-exception-caught
+            log.debug2("Caught unexpected error: %s", err, exc_info=True)
+        return False
+
+
+def _grpc_health_probe(
+    port: int,
+    tls_key: Optional[str],
+    tls_cert: Optional[str],
+    client_ca: Optional[str],
+) -> bool:
+    """Probe the grpc server
+
+    Since the gRPC server trusts its own cert for client verification, we can
+    make a valid health probe against the running server regardless of (m)TLS
+    config.
+
+    Args:
+        port (int): The port that the gRPC server is serving on
+        tls_key (Optional[str]): Body or path to the TLS key file if TLS/mTLS
+            enabled
+        tls_cert (Optional[str]): Body or path to the TLS cert file if TLS/mTLS
+            enabled
+        client_ca (Optional[str]): The client ca cert that the server is using
+            for mutual client auth
+
+    Returns:
+        healthy (bool): True if all servers are healthy, False otherwise
+    """
+    # NOTE: Local imports for optional dependency
+    with alog.ContextTimer(log.debug2, "Done with local grpc imports: "):
+
+        # Third Party
+        from grpc_health.v1 import (  # pylint: disable=import-outside-toplevel
+            health_pb2,
+            health_pb2_grpc,
+        )
+        import grpc  # pylint: disable=import-outside-toplevel
+
+        # Local
+        from .grpc_server import (  # pylint: disable=import-outside-toplevel
+            RuntimeGRPCServer,
+        )
+
+    hostname = f"localhost:{port}"
+    if tls_key and tls_cert:
+        tls_server_key = bytes(RuntimeGRPCServer._load_secret(tls_key), "utf-8")
+        tls_server_cert = bytes(RuntimeGRPCServer._load_secret(tls_cert), "utf-8")
+        if client_ca:
+            log.debug("Probing mTLS gRPC server")
+            credentials = grpc.ssl_channel_credentials(
+                root_certificates=tls_server_cert,
+                private_key=tls_server_key,
+                certificate_chain=tls_server_cert,
+            )
+        else:
+            log.debug("Probing TLS gRPC server")
+            credentials = grpc.ssl_channel_credentials(
+                root_certificates=tls_server_cert,
+            )
+        channel = grpc.secure_channel(hostname, credentials=credentials)
+    else:
+        log.debug("Probing INSECURE gRPC server")
+        channel = grpc.insecure_channel(hostname)
```

Comment:
would creating a new channel with every call be slow?

Reply:
Yes-ish. This will be run as a standalone process, so we can't hold anything in memory between calls. I do think the tls files should be cached though

Reply:
it sure would!

But like the files bit, this is a binary meant to be run once on each check. Dunno how you cache a `channel` across multiple processes like that :(